Manage access requests programmatically

Wauplin opened the pull request:
Solves https://github.com/huggingface/huggingface_hub/issues/1535.

Let's add official support to programmatically manage access requests for gated repos.

The API defines a `AccessRequest` dataclass with username, email, status, timestamp and fields + 7 endpoints. I reused as much as possible the vocabulary used in the UI. We have:
- `list_pending_access_requests` => returns "pending" list
- `list_accepted_access_requests` => returns "accepted" list
- `list_rejected_access_requests` => returns "rejected" list
- `cancel_access_request` => moves request to "pending" list
- `accept_access_request` => moves request to "accepted" list  
- `reject_access_request` => moves request to "rejected" list
- `grant_access` => add a user to "accepted" list (the user doesn't have to request access first)

Methods docstrings are documented with examples. I don't think it's worth creating a guide dedicated to it in `huggingface_hub` as it would be quite redundant with the Hub docs already. I suggest to update the Hub docs instead to mention the `huggingface_hub` implementation but that's it. 

TODO:
- [ ] update https://huggingface.co/docs/hub/models-gated and https://huggingface.co/docs/hub/datasets-gated once this PR is shipped

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/huggingface_hub/pr_1905). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

osanseviero commented on src/huggingface_hub/hf_api.py:
```diff
@@ -7655,6 +7684,368 @@ def delete_collection_item(
+
+        A rejected request means the user has requested access to the repo and the request has been explicitly rejected
+        by a repo owner (either you or another user from your organization). The user cannot download any file of the
+        repo and cannot access the community tab. Rejected requests can be accepted or cancelled at any time using
```
> and cannot access the community tab

Are you sure about this? I thought all users had access to community tab

Wauplin replied:
Wow sorry, don't know from where I invented that. I think I'm hallucinating facts :face_with_spiral_eyes:

Wauplin replied:
Changed in https://github.com/huggingface/huggingface_hub/pull/1905/commits/31a35dab75e7bfb4b75c26a0d904df887b007780

julien-c replied:
you used a LLM copilot maybe

osanseviero replied:
LM stands for Lucain Model

osanseviero reviewed (approved):
Nice!

Wauplin commented on src/huggingface_hub/hf_api.py:
```diff
@@ -324,6 +324,35 @@ def __post_init__(self):
+            Timestamp of the request.
+        status (`Literal["pending", "accepted", "rejected"]`):
+            Status of the request. Can be one of `["pending", "accepted", "rejected"]`.
+        fields (`Dict[str, str]`, *optional*):
```
yeah, let's not assume a type. Changed to `Dict[str, Any]`

LysandreJik commented on src/huggingface_hub/hf_api.py:
```diff
@@ -7655,6 +7684,367 @@ def delete_collection_item(
+                A valid authentication token (see https://huggingface.co/settings/token).
+
+        Raises:
+            - `HTTPError`: HTTP 400 if the repo is not gated.
+            - `HTTPError`: HTTP 404 if the user does not exist on the Hub.
+            - `HTTPError`: HTTP 404 if the user access request cannot be found.
+            - `HTTPError`: HTTP 404 if the user access request is already in the accepted list.
```
403 if auth error :stuck_out_tongue: (but maybe that's out of scope of what you're aiming for here)

LysandreJik commented on tests/test_hf_api.py:
```diff
@@ -3668,3 +3671,85 @@ def test_collection_items(self) -> None:
+        self._api.delete_repo(self.repo_id)
+        return super().tearDown()
+
+    def test_access_requests_normal_usage(self) -> None:
```
Intensive test :ok_hand:

LysandreJik reviewed (approved):
That's awesome! Very cool API.

Is it worth mentioning what kind of permissions they'll need on the repo to have access to the accepted/pending/rejected requests (is it read/write/admin?)? Especially given the example is Llama-2-7b.

Wauplin commented:
Thanks both for the review!

I have addressed your comments. @LysandreJik regarding roles in organization, the user must have "write" or "admin" role. I have updated the docstrings accordingly. (fyi, role `contributor` also have access to these settings on repos for which they have write access. It's a corner case so I haven't documented it).

I'll wait for the CI to pass and then merge this!

RonanKMcGovern commented:
@lappemic @Wauplin has something changed here in the last days? After April 18th, my automations querying access seem to have stopped working. Curl requests like this:
```
curl -X GET \             
  'https://huggingface.co/api/repos/Trelis/Mistral-7B-Instruct-v0.1-Summarize-64k/user-access-request/pending' \
  -H 'Authorization: Bearer TOKEN'
```
seem not to be working any more

Wauplin commented:
> has something changed here in the last days?

@RonanKMcGovern shouldn't be the case no.

If you are getting errors, could you open a new ticket and provide more context (requests/script + error message + stacktrace if applicable). Thanks in advance!

RonanKMcGovern commented:
Ah very sorry @Wauplin , I just realised I rolled my HF token after a live stream last week without realising it was being used in some automations. Thanks for the quick reply.

Wauplin commented:
Glad your issue's solved then!
